src/transformers/models/seggpt/image_processing_seggpt.py

```diff
@@ -348,7 +348,10 @@ def _preprocess_step(
         size = size if size is not None else self.size
         size_dict = get_size_dict(size)
 
-        images = make_list_of_images(images)
+        try:
+            images = make_list_of_images(images)
+        except ValueError:
+            images = make_list_of_images(images, expected_ndims=2)
```

Comment:
No - we shouldn't use try / expect here. In fact, I don't think this call is robust if the input image has two dimensions as that's not currently compatible with many of the image transformations. 

It would be better to have two separate methods, `_preprocess_image` and `_preprocess_mask`, each of which call `_preprocess`. `_preprocess_mask` can then use `mask_to_rgb`, before calling `_preprocess` and we can remove this step entirely for when we process images.

Reply:
`mask_to_rgb` currently expects a list of numpy arrays and this conversion happens on the `_preprocess_step` method with `make_list_of_images` and `to_numpy_array`. The problem is that we are currently accepting/expecting 2D (semantic maps) and 3D (if already converted to RGB with some logic) images for mask prompt. Maybe an easy fix would be to allow the input of the expected dimension for the mask something like `prompt_mask_dimension` and then pass that to `make_list_of_images`, wdyt?

Reply:
> mask_to_rgb currently expects a list of numpy arrays 

I don't think this is true - `mask_to_rgb` expects a single image

> Maybe an easy fix would be to allow the input of the expected dimension for the mask something like prompt_mask_dimension and then pass that to make_list_of_images, wdyt?

This isn't ideal. It would be better for `mask_to_rgb` accept masks with and without channels and convert as necessary to RGB for both cases.

Reply:
Ops my bad, `mask_to_rgb` expects a single image you're right. 

But, still, the current problem lies on the `make_list_of_images` even if we split `_preprocess` to `_preprocess_image` and `_preprocess_mask` this problem would still be there and since `make_list_of_images` takes an `ImageInput` as arg which can be either `PIL`, `numpy`, ... and even if we had this logic to parse the input somewhere else it would still be ambiguous if a 3D mask is a batch of zero channel images or a single RGB mask (for inputs other than `PIL`)

Reply:
OK, I see, so you're saying that the difficulty comes from the face the passed in masks can be `(B, H, W)` or `(B, C, H, W)`? The current approach wouldn't fix this, as it would still make a batch of `(B, H, W)` masks a list of a `[batch]` instead of `[mask_0, mask_1, mask_2, ... ]` 

I would just add a custom `make_list_of_images` for this model, which tasks the images and masks and makes a list of the masks based on the batch size of the input images

Reply:
@amyeroberts Yeah, but there's another issue to that in the sense that if the image is 3D it can either be `(B, H, W)` batched semantic map or `(C, H, W)` RGB image. This will only be a problem if we have `ChannelDimension.First` and `B=3`

Therefore, even creating a custom `make_list_of_images` (or `make_list_of_masks` as the problem is in the masks) we'll have ambiguity in the case where `image.ndim==3`.

This is the portion of `make_list_of_images` that raises problems:

```python
if is_valid_image(images):
        if images.ndim == expected_ndims + 1:
            # Batch of images
            images = list(images)
        elif images.ndim == expected_ndims:
            # Single image
            images = [images]
        else:
            raise ValueError(
                f"Invalid image shape. Expected either {expected_ndims + 1} or {expected_ndims} dimensions, but got"
                f" {images.ndim} dimensions."
            )
        return images
```

At this point the images are either `np.ndarray` or a `torch.tensor` in the context of `SegGpt`. 

I've created a new function called `make_list_of_masks` here https://github.com/huggingface/transformers/pull/29550/commits/7ea7a41b02cc8393f1c265b66c2f7091f6f9e0a4 and deal with that, but had to make an assumption that if `images` is 3D then if the last of first dim is 3 it's a single RGB image.